File: src/pipeline/claudeVisionCritic.ts
import Anthropic from "@anthropic-ai/sdk";
import type { VisionCritic } from "./types";
import type { Fix, IssueKind, ReviewIssue, SceneState, Severity, Vec3 } from "../scene/schema";
import { loadConfig } from "../config";

// Real vision critic backed by Claude vision (@anthropic-ai/sdk). Opus 4.8 sees the
// rendered screenshot plus the layout JSON and reports placement problems through a
// forced tool call, which we validate into typed ReviewIssue[] tagged source:"vision".
const MODEL = "claude-opus-4-8";
const MAX_TOKENS = 4096;

// The kinds a vision pass can detect (geometry handles the rest); each maps to one fix op.
const ISSUE_KINDS: IssueKind[] = [
  "overlap",
  "floating",
  "wrong_facing",
  "too_big",
  "too_small",
  "out_of_bounds",
];
const SEVERITIES: Severity[] = ["low", "medium", "high"];
const FIX_OPS = ["move", "rotate", "resize"] as const;
type ImageMediaType = "image/jpeg" | "image/png" | "image/gif" | "image/webp";

const SYSTEM_PROMPT = [
  "You are a meticulous reviewer of rendered 3D interior scenes.",
  "You receive a screenshot of the current layout and the layout JSON (each object's id,",
  "label, size, and transform). Compare the image against the layout and report concrete,",
  "visible placement problems by calling the emit_review_issues tool exactly once.",
  "",
  "Each issue must reference an existing object id and carry exactly one concrete fix:",
  "- overlap / out_of_bounds / floating  -> op 'move', delta [dx, dy, dz] in meters.",
  "- wrong_facing                        -> op 'rotate', rotationYDeg degrees to add.",
  "- too_big / too_small                 -> op 'resize', scaleFactor (>1 enlarges, <1 shrinks).",
  "Report only real problems you can see. If the scene looks correct, return an empty list.",
].join("\n");

const VEC3_SCHEMA = { type: "array", items: { type: "number" }, minItems: 3, maxItems: 3 } as const;

const REVIEW_TOOL: Anthropic.Tool = {
  name: "emit_review_issues",
  description: "Return the placement problems found in the rendered scene (empty if none).",
  input_schema: {
    type: "object",
    additionalProperties: false,
    required: ["issues"],
    properties: {
      issues: {
        type: "array",
        description: "zero or more concrete issues; empty when the scene looks correct",
        items: {
          type: "object",
          additionalProperties: false,
          required: ["objectId", "kind", "severity", "description", "fix"],
          properties: {
            objectId: { type: "string", description: "id of the offending object; must exist in the layout" },
            kind: { type: "string", enum: ISSUE_KINDS },
            severity: { type: "string", enum: SEVERITIES },
            description: { type: "string", description: "one concrete sentence describing the visible problem" },
            fix: {
              type: "object",
              additionalProperties: false,
              required: ["op"],
              properties: {
                op: { type: "string", enum: FIX_OPS },
                delta: { ...VEC3_SCHEMA, description: "move op: meters added to position [dx, dy, dz]" },
                rotationYDeg: { type: "number", description: "rotate op: degrees added to current yaw" },
                scaleFactor: { type: "number", description: "resize op: multiplier on current scale (>1 bigger)" },
              },
            },
          },
        },
      },
    },
  },
};

export const claudeVisionCritic: VisionCritic = {
  async review(input: { scene: SceneState; screenshotDataUrl: string }): Promise<ReviewIssue[]> {
    const { scene, screenshotDataUrl } = input;
    const image = parseScreenshot(screenshotDataUrl);
    const knownIds = new Set(scene.objects.map((o) => o.id));

    const { anthropicApiKey } = loadConfig();
    const client = new Anthropic({ apiKey: anthropicApiKey });

    const response = await client.messages.create({
      model: MODEL,
      max_tokens: MAX_TOKENS,
      system: SYSTEM_PROMPT,
      tools: [REVIEW_TOOL],
      tool_choice: { type: "tool", name: REVIEW_TOOL.name },
      messages: [
        {
          role: "user",
          content: [
            { type: "image", source: { type: "base64", media_type: image.mediaType, data: image.data } },
            { type: "text", text: `Layout JSON:\n${layoutSummary(scene)}` },
          ],
        },
      ],
    });

    const toolUse = response.content.find((block) => block.type === "tool_use");
    if (!toolUse || toolUse.type !== "tool_use") {
      throw new Error(
        `claudeVisionCritic: model did not return a tool_use block (stop_reason=${response.stop_reason})`,
      );
    }

    const result = asRecord(toolUse.input, "review result");
    if (!Array.isArray(result.issues)) {
      throw new Error(`claudeVisionCritic: expected issues[] array, got ${describe(result.issues)}`);
    }

    return result.issues.map((raw, i) => {
      const issue = asRecord(raw, `issues[${i}]`);
      const objectId = asNonEmptyString(issue.objectId, `issues[${i}].objectId`);
      if (!knownIds.has(objectId)) {
        throw new Error(`claudeVisionCritic: issues[${i}].objectId "${objectId}" is not in the scene`);
      }
      return {
        objectId,
        kind: asEnum(issue.kind, ISSUE_KINDS, `issues[${i}].kind`),
        severity: asEnum(issue.severity, SEVERITIES, `issues[${i}].severity`),
        description: asNonEmptyString(issue.description, `issues[${i}].description`),
        fix: parseFix(issue.fix, `issues[${i}].fix`),
        source: "vision",
      };
    });
  },
};

// Compact, model-facing view of the scene so the critic can map pixels to object ids.
function layoutSummary(scene: SceneState): string {
  return JSON.stringify(
    {
      room: scene.room,
      pass: scene.pass,
      objects: scene.objects.map((o) => ({
        id: o.id,
        label: o.label,
        approxSize: o.approxSize,
        position: o.transform.position,
        rotationYDeg: o.transform.rotationYDeg,
        scale: o.transform.scale,
        status: o.status,
      })),
    },
    null,
    2,
  );
}

function parseScreenshot(dataUrl: string): { mediaType: ImageMediaType; data: string } {
  const match = /^data:(image\/(?:png|jpe?g|gif|webp));base64,(.+)$/s.exec(dataUrl.trim());
  if (!match) {
    throw new Error("claudeVisionCritic: screenshotDataUrl must be a base64-encoded image data URL");
  }
  const declared = match[1];
  const mediaType: ImageMediaType = declared === "image/jpg" ? "image/jpeg" : (declared as ImageMediaType);
  return { mediaType, data: match[2] };
}

// --- validation helpers: the model's tool input is untrusted, so narrow loudly ---

function parseFix(value: unknown, ctx: string): Fix {
  const raw = asRecord(value, ctx);
  const op = asEnum(raw.op, FIX_OPS, `${ctx}.op`);
  if (op === "move") return { op, delta: asVec3(raw.delta, `${ctx}.delta`) };
  if (op === "rotate") return { op, rotationYDeg: asNumber(raw.rotationYDeg, `${ctx}.rotationYDeg`) };
  return { op, scaleFactor: asNumber(raw.scaleFactor, `${ctx}.scaleFactor`) };
}

function asEnum<T extends string>(value: unknown, allowed: readonly T[], ctx: string): T {
  if (typeof value !== "string" || !allowed.includes(value as T)) {
    throw new Error(
      `claudeVisionCritic: expected one of [${allowed.join(", ")}] at ${ctx}, got ${describe(value)}`,
    );
  }
  return value as T;
}

function asRecord(value: unknown, ctx: string): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new Error(`claudeVisionCritic: expected an object at ${ctx}, got ${describe(value)}`);
  }
  return value as Record<string, unknown>;
}

function asNumber(value: unknown, ctx: string): number {
  if (typeof value !== "number" || !Number.isFinite(value)) {
    throw new Error(`claudeVisionCritic: expected a finite number at ${ctx}, got ${describe(value)}`);
  }
  return value;
}

function asNonEmptyString(value: unknown, ctx: string): string {
  if (typeof value !== "string" || value.trim().length === 0) {
    throw new Error(`claudeVisionCritic: expected a non-empty string at ${ctx}, got ${describe(value)}`);
  }
  return value;
}

function asVec3(value: unknown, ctx: string): Vec3 {
  if (!Array.isArray(value) || value.length !== 3) {
    throw new Error(`claudeVisionCritic: expected a [x, y, z] array at ${ctx}, got ${describe(value)}`);
  }
  return [
    asNumber(value[0], `${ctx}[0]`),
    asNumber(value[1], `${ctx}[1]`),
    asNumber(value[2], `${ctx}[2]`),
  ];
}

function describe(value: unknown): string {
  if (value === null) return "null";
  if (Array.isArray(value)) return `array(len ${value.length})`;
  return typeof value;
}

File: backend/src/pipeline/claude_vision_critic.py
"""Claude vision critic for rendered Dream3D scenes."""

import base64
import json
import math
import re
import shutil
import tempfile
from pathlib import Path
from typing import Any

from llm.claude_cli import run_claude
from pipeline.prompts import load
from scene.schema import Fix, ReviewIssue, SceneState, Vec3

ISSUE_KINDS = (
    "overlap",
    "floating",
    "wrong_facing",
    "too_big",
    "too_small",
    "out_of_bounds",
)
SEVERITIES = ("low", "medium", "high")
FIX_OPS = ("move", "rotate", "resize")

EXT_BY_MEDIA = {
    "image/jpeg": "jpg",
    "image/png": "png",
    "image/gif": "gif",
    "image/webp": "webp",
}

SYSTEM_PROMPT = load("vision_critic_system.md")

VEC3_SCHEMA = {
    "type": "array",
    "items": {"type": "number"},
    "minItems": 3,
    "maxItems": 3,
}

REVIEW_SCHEMA = {
    "type": "object",
    "additionalProperties": False,
    "required": ["issues"],
    "properties": {
        "issues": {
            "type": "array",
            "description": "zero or more concrete issues; empty when the scene looks correct",
            "items": {
                "type": "object",
                "additionalProperties": False,
                "required": ["objectId", "kind", "severity", "description", "fix"],
                "properties": {
                    "objectId": {
                        "type": "string",
                        "description": "id of the offending object; must exist in the layout",
                    },
                    "kind": {"type": "string", "enum": ISSUE_KINDS},
                    "severity": {"type": "string", "enum": SEVERITIES},
                    "description": {
                        "type": "string",
                        "description": "one concrete sentence describing the visible problem",
                    },
                    "fix": {
                        "type": "object",
                        "additionalProperties": False,
                        "required": ["op"],
                        "properties": {
                            "op": {"type": "string", "enum": FIX_OPS},
                            "delta": {
                                **VEC3_SCHEMA,
                                "description": "move op: meters added to position [dx, dy, dz]",
                            },
                            "rotationYDeg": {
                                "type": "number",
                                "description": "rotate op: degrees added to current yaw",
                            },
                            "scaleFactor": {
                                "type": "number",
                                "description": "resize op: multiplier on current scale (>1 bigger)",
                            },
                        },
                    },
                },
            },
        },
    },
}

DATA_URL_RE = re.compile(r"^data:(image/(?:png|jpe?g|gif|webp));base64,(.+)$", re.S)


async def review(scene: SceneState, views: list[dict[str, str]]) -> list[ReviewIssue]:
    if len(views) == 0:
        raise ValueError("claudeVisionCritic: review requires at least one rendered view")
    known_ids = {obj.id for obj in scene.objects}

    directory = Path(tempfile.mkdtemp(prefix="dream3d-critic-"))
    try:
        image_paths: list[str] = []
        for view in views:
            name = view["name"]
            image = _parse_data_url(view["dataUrl"], name)
            image_path = directory / f"{name}.{EXT_BY_MEDIA[image['media_type']]}"
            image_path.write_bytes(base64.b64decode(image["data"], validate=True))
            image_paths.append(str(image_path))

        prompt = "\n".join(
            [
                SYSTEM_PROMPT,
                "",
                f"The {len(views)} image(s) are labeled camera angles of the SAME scene ({', '.join(view['name'] for view in views)}) —",
                "the same objects seen from different positions, not different scenes. Report each object AT MOST ONCE",
                "across all angles, citing the angle where the problem is clearest.",
                "",
                f"Layout JSON:\n{layout_summary(scene)}",
                "",
                "Respond with ONLY the JSON object — no prose, no markdown code fences. It must match this schema:",
                json.dumps(REVIEW_SCHEMA, indent=2),
            ]
        )
        text = run_claude(prompt, image_paths=image_paths, caller="vision-critic")
    finally:
        shutil.rmtree(directory)

    result = _as_record(_parse_json(text), "review result")
    issues = result.get("issues")
    if not isinstance(issues, list):
        raise ValueError(
            f"claudeVisionCritic: expected issues[] array, got {_describe(issues)}"
        )

    parsed: list[ReviewIssue] = []
    for i, raw in enumerate(issues):
        issue = _as_record(raw, f"issues[{i}]")
        object_id = _as_non_empty_string(issue.get("objectId"), f"issues[{i}].objectId")
        if object_id not in known_ids:
            raise ValueError(
                f'claudeVisionCritic: issues[{i}].objectId "{object_id}" is not in the scene'
            )
        parsed.append(
            ReviewIssue(
                objectId=object_id,
                kind=_as_enum(issue.get("kind"), ISSUE_KINDS, f"issues[{i}].kind"),
                severity=_as_enum(
                    issue.get("severity"), SEVERITIES, f"issues[{i}].severity"
                ),
                description=_as_non_empty_string(
                    issue.get("description"), f"issues[{i}].description"
                ),
                fix=_parse_fix(issue.get("fix"), f"issues[{i}].fix"),
                source="vision",
            )
        )
    return parsed


def layout_summary(scene: SceneState) -> str:
    return json.dumps(
        {
            "room": scene.room.model_dump(),
            "pass": scene.pass_,
            "objects": [
                {
                    "id": obj.id,
                    "label": obj.label,
                    "approxSize": obj.approx_size,
                    "position": obj.transform.position,
                    "rotationYDeg": obj.transform.rotation_y_deg,
                    "scale": obj.transform.scale,
                    "status": obj.status,
                }
                for obj in scene.objects
            ],
        },
        indent=2,
    )


def _parse_data_url(data_url: str, view_name: str) -> dict[str, str]:
    match = DATA_URL_RE.match(data_url.strip())
    if not match:
        raise ValueError(
            f'claudeVisionCritic: view "{view_name}" dataUrl must be a base64-encoded image data URL'
        )
    declared = match.group(1)
    media_type = "image/jpeg" if declared == "image/jpg" else declared
    return {"media_type": media_type, "data": match.group(2)}


def _parse_json(text: str) -> Any:
    body = text.strip()
    fence = re.search(r"```(?:json)?\s*\n?([\s\S]*?)```", body)
    if fence:
        body = fence.group(1).strip()
    try:
        return json.loads(body)
    except json.JSONDecodeError as exc:
        salvaged = _slice_outermost_json(body)
        if salvaged is not None:
            try:
                return json.loads(salvaged)
            except json.JSONDecodeError as salvage_exc:
                exc.add_note(f"salvaged JSON parse also failed: {salvage_exc}")
        raise ValueError(
            f"claudeVisionCritic: model output was not valid JSON: {body[:1000]}"
        ) from exc


def _slice_outermost_json(body: str) -> str | None:
    open_brace = body.find("{")
    if open_brace != -1:
        close_brace = body.rfind("}")
        if close_brace > open_brace:
            return body[open_brace : close_brace + 1]
        return None
    open_bracket = body.find("[")
    if open_bracket != -1:
        close_bracket = body.rfind("]")
        if close_bracket > open_bracket:
            return body[open_bracket : close_bracket + 1]
    return None


def _parse_fix(value: Any, ctx: str) -> Fix:
    raw = _as_record(value, ctx)
    op = raw.get("op")
    if op == "move":
        return Fix(op="move", delta=_as_vec3(raw.get("delta"), f"{ctx}.delta"))
    if op == "rotate":
        return Fix(
            op="rotate",
            rotationYDeg=_as_number(raw.get("rotationYDeg"), f"{ctx}.rotationYDeg"),
        )
    if op == "resize":
        return Fix(
            op="resize",
            scaleFactor=_as_number(raw.get("scaleFactor"), f"{ctx}.scaleFactor"),
        )
    raise ValueError(
        f"claudeVisionCritic: expected one of [{', '.join(FIX_OPS)}] at {ctx}.op, got {_describe(op)}"
    )


def _as_enum(value: Any, allowed: tuple[str, ...], ctx: str) -> str:
    if not isinstance(value, str) or value not in allowed:
        raise ValueError(
            f"claudeVisionCritic: expected one of [{', '.join(allowed)}] at {ctx}, got {_describe(value)}"
        )
    return value


def _as_record(value: Any, ctx: str) -> dict[str, Any]:
    if not isinstance(value, dict):
        raise ValueError(
            f"claudeVisionCritic: expected an object at {ctx}, got {_describe(value)}"
        )
    return value


def _as_number(value: Any, ctx: str) -> float:
    if (
        isinstance(value, bool)
        or not isinstance(value, (int, float))
        or not math.isfinite(value)
    ):
        raise ValueError(
            f"claudeVisionCritic: expected a finite number at {ctx}, got {_describe(value)}"
        )
    return float(value)


def _as_non_empty_string(value: Any, ctx: str) -> str:
    if not isinstance(value, str) or len(value.strip()) == 0:
        raise ValueError(
            f"claudeVisionCritic: expected a non-empty string at {ctx}, got {_describe(value)}"
        )
    return value


def _as_vec3(value: Any, ctx: str) -> Vec3:
    if not isinstance(value, list) or len(value) != 3:
        raise ValueError(
            f"claudeVisionCritic: expected a [x, y, z] array at {ctx}, got {_describe(value)}"
        )
    return (
        _as_number(value[0], f"{ctx}[0]"),
        _as_number(value[1], f"{ctx}[1]"),
        _as_number(value[2], f"{ctx}[2]"),
    )


def _describe(value: Any) -> str:
    if value is None:
        return "null"
    if isinstance(value, list):
        return f"array(len {len(value)})"
    if isinstance(value, bool):
        return "boolean"
    if isinstance(value, str):
        return "string"
    if isinstance(value, (int, float)):
        return "number"
    return type(value).__name__
